app/code/Magento/Inventory/Model/SourceRepository.php

```diff
@@ -0,0 +1,190 @@
+     */
+    private $resourceSource;
+
+    /**
+     * @var ResourceSourceCarrierLink
+     */
+    private $resourceSourceCarrierLink;
+
+    /**
+     * @var SourceInterfaceFactory
+     */
+    private $sourceFactory;
+
+    /**
+     * @var CollectionProcessorInterface
+     */
+    private $collectionProcessor;
+
+    /**
+     * @var CollectionFactory
+     */
+    private $collectionFactory;
+
+    /**
+     * @var CarrierLinkCollectionFactory
+     */
+    private $carrierLinkCollectionFactory;
+
+    /**
+     * @var SourceSearchResultsInterfaceFactory
+     */
+    private $sourceSearchResultsFactory;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * SourceRepository constructor.
+     * @param ResourceSource $resourceSource
+     * @param ResourceSourceCarrierLink $resourceSourceCarrierLink
+     * @param SourceInterfaceFactory $sourceFactory
+     * @param CollectionProcessorInterface $collectionProcessor
+     * @param CollectionFactory $collectionFactory
+     * @param CarrierLinkCollectionFactory $carrierLinkCollectionFactory
+     * @param SourceSearchResultsInterfaceFactory $sourceSearchResultsFactory
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
+        ResourceSource $resourceSource,
+        ResourceSourceCarrierLink $resourceSourceCarrierLink,
+        SourceInterfaceFactory $sourceFactory,
+        CollectionProcessorInterface $collectionProcessor,
+        CollectionFactory $collectionFactory,
+        CarrierLinkCollectionFactory $carrierLinkCollectionFactory,
+        SourceSearchResultsInterfaceFactory $sourceSearchResultsFactory,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        LoggerInterface $logger
+    ) {
+        $this->resourceSource = $resourceSource;
+        $this->resourceSourceCarrierLink = $resourceSourceCarrierLink;
+        $this->sourceFactory = $sourceFactory;
+        $this->collectionProcessor = $collectionProcessor;
+        $this->collectionFactory = $collectionFactory;
+        $this->carrierLinkCollectionFactory = $carrierLinkCollectionFactory;
+        $this->sourceSearchResultsFactory = $sourceSearchResultsFactory;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function save(SourceInterface $source)
+    {
+        try {
+            /** @var SourceInterface|AbstractModel $source */
+            $this->resourceSource->save($source);
+
+            /** @var SourceCarrierLinkInterface|AbstractModel $carrierLink */
+            foreach ($source->getCarrierLinks() as $carrierLink) {
+                $carrierLink->setData(SourceInterface::SOURCE_ID, $source->getSourceId());
+                $this->resourceSourceCarrierLink->save($carrierLink);
```

Comment:
it's better to use Bulk insert/delete. Use Bulk inserts as methods of the resource model.
It will speed up performance